peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+``TypeGuard`` are occasionally useful, and we do not propose to deprecate or remove it. In the long
+run, most users should use ``TypeNarrower``, and ``TypeGuard`` should be reserved for rare cases
+where its behavior is specifically desired.
+
+
+Specification
+=============
+
+A new special form, ``TypeNarrower``, is added to the :py:mod:`typing`
+module. Its usage, behavior, and runtime implementation are similar to
+those of :py:data:`typing.TypeGuard`.
+
+It accepts a single
+argument and can be used as the return type of a function. A function annotated as returning a
+``TypeNarrower`` is called a type narrowing function. Type narrowing functions must return ``bool``
+values, and the type checker should verify that all return paths return
+``bool``. However, ``TypeNarrower`` is not a subtype of ``bool``.
+The type ``Callable[..., TypeNarrower[int]]`` is not assignable to
+``Callable[..., bool]`` or ``Callable[..., TypeGuard[int]]``, and vice versa.
+
+Type narrowing functions must accept at least one positional argument. The type
+narrowing behavior is applied to the first positional argument passed to
+the function. The function may accept additional arguments, but they are
+not affected by type narrowing.
+
+Type narrowing behavior
+-----------------------
+
+To specify the behavior of ``TypeNarrower``, we use the following terminology:
+
+* I = ``TypeNarrower`` input type
+* R = ``TypeNarrower`` return type
+* A = Type of argument passed to type narrowing function (pre-narrowed)
+* NP = Narrowed type (positive; used when ``TypeNarrower`` returned True)
+* NN = Narrowed type (negative; used when ``TypeNarrower`` returned False)
+
+.. code-block:: python
+
+    def narrower(x: I) -> TypeNarrower[R]: ...
+
+    def func1(val: A):
+        if narrower(val):
+            assert_type(val, NP)
+        else:
+            assert_type(val, NN)
+
+The return type ``R`` must be :ref:`consistent with <pep-483-gradual-typing>` ``I``. The type checker should
+emit an error if this condition is not met.
+
+Formally, type *NP* should be narrowed to :math:`A \land R`, the intersection of *A* and *R*, and type *NN* should be narrowed to
+:math:`A \land \neg R`, the intersection of *A* and the complement of *R*.
+In practice, the theoretic types for strict type guards cannot be expressed
+precisely in the Python type system. Type checkers should fall back on
+practical approximations of these types. As a rule of thumb, a type checker
+should use the same type narrowing logic -- and get results that are consistent
+with -- its handling of :py:func:`isinstance`. This guidance allows for changes and
+improvements if the type system is extended in the future.
+
+In particular, if the argument type ``A`` is a union type, ``NP`` may
+be narrowed to those elements of the union that are consistent with ``R``,
+and ``NN`` may be narrowed to those elements of the union that are
+not subtypes of ``R``.
```

Comment:
```suggestion
not consistent subtypes of ``R``.
```

Reply:
I did write "subtypes" here on purpose. My understanding is that Any is a consistent subtype of any concrete type T, but not a subtype. If we start with a type `A | Any` and have a narrower that narrows to `A`, then in the else case the type should be narrowed to `Any`, not Never.

Reply:
I think the right wording is that NP includes those union elements that are subtypes of R, and NN includes those that are not subtypes of R. I will change the text to say that.

Reply:
Thinking about it more, this is still wrong in the presence of generics. Consider the case where `A` is `list[Any] | set[str]` and `R` is `Iterable[str]`. Then I think `NP` would ideally be `list[str] | set[str]` and `NN` would be `list[Any]`, but this specification would dictate that `NP` is just `set[str]`.

I think this specification only holds for literal types. I'll reword it to say it's an example and limited to literal types. It would be a mistake to attempt a general specification for intersection types here.

Reply:
I'm actually not sure what you mean by "literal types" here? My initial reading of that term is that it includes only types using the `Literal` special form. If that is what you mean, it seems like an oddly specific example that might cause more confusion than it clears up ("what's so special about `Literal`?"). If that's not what you mean, then I think we need a clearer term.

I don't think certain kinds of types are special; from the examples given so far, I think the correct specification is asymmetrical: NP are those elements of the union that are consistent subtypes of R, and NN are those elements of the union that are not subtypes of R. In other words, `Any` can always be ruled in and never ruled out. (If I'm not mistaken, this is the wording you originally had, so I'm sorry that I confused it with a not-thought-through comment!)

Another approach (since we are not trying to provide a full spec here) might be to eliminate this paragraph and provide just a list of concrete examples.

Reply:
I did mean `Literal` only. Those types have the advantage that they cannot be subclassed or be generic, which makes life simpler.

I don't think your spec behaves the way I'd want a type checker to behave. For example, if `A` is `Any | str` and `R` is `str`, I would want `NP` to be `str`, not `Any | str`.

Still, you're right that the example becomes oddly specific if it's just about literals. I'll think about another rewording.

Reply:
> I don't think your spec behaves the way I'd want a type checker to behave.

The spec, as it's currently written, doesn't dictate the exact behavior. It provides the type-theoretically-correct formulation, and it recommends that type checkers employ similar logic that they do for `isinstance` type narrowing. I think that's the best we can do here. 

Rich Chiodo and I did an in-depth analysis and presentation at a typing meetup on this topic. If you didn't attend that meeting or would like to refresh your memory, check out the presentation [here](https://www.dropbox.com/scl/fi/8rdvtoiwxf515ln61yjxh/Strict-Type-Guards.pdf?rlkey=gn8a4dcq8tscv8n4ebvmorc9f&dl=0).

We should probably choose an example in the spec that is straightforward, non-controversial, and likely to generate the same results for all type checkers (based on their current `isinstance` narrowing behaviors).

Reply:
I don't think any spec framed solely in terms of inclusion/exclusion of items in a union can handle the generic examples, where we'd ideally want to refine the type parameters of an element of the original union.

I think the formal spec in terms of intersections (and the parallel to `isinstance`) is perhaps the only spec that needs to be provided here. A fuller specification of how type checkers approximate intersections might be useful, but seems out of scope for this PEP. Replacing this paragraph with a list of concrete "desirable examples" seems like it provides a practical illustration of the previous paragraph, without trying to be a full intersection spec. I think that will be clearer than trying to provide a more general "example" in prose.

Reply:
Thanks Eric, the presentation was very helpful. Perhaps it should be linked from the PEP?

I dropped the paragraph this discussion is about. The previous paragraph already gives a spec in terms of intersection semantics (mirroring the spec in PEP 724) and the next section contains examples.